contrib/insights/src/column_statistics.h

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You
+ * may not use this file except in compliance with the License. A copy of
+ * the License is located at
+ *
+ *      http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is
+ * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+ * ANY KIND, either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+#pragma once
+
+#include "mlio.h"
+#include "hll/hyperloglog.h"
+
+namespace insights {
+namespace detail {
+
+static const uint8_t CARDINALITY_HLL_SIZE = 16;
+class MLIO_API column_analysis {
+    public:
+
+    column_analysis(std::string name) 
+        : rows_seen(0),
+          numeric_mean(0.0), 
+          numeric_count(0), 
+          numeric_nan_count(0), 
+          numeric_min(std::numeric_limits<double>::quiet_NaN()),
+          numeric_max(std::numeric_limits<double>::quiet_NaN()),
+          string_min_length(std::numeric_limits<size_t>::max()),
+          string_min_length_not_empty(std::numeric_limits<size_t>::max()),
+          string_max_length(0),
+          string_empty_count(0),
+          string_only_whitespace_count(0),
+          string_null_like_count(0),
+          string_captured_unique_values_overflowed(false),
+          example_value("")
+    { column_name = name; 
+      string_cardinality_estimator = hll::HyperLogLog(CARDINALITY_HLL_SIZE); 
+    }
+
+    size_t string_cardinality() const {
+      return static_cast<size_t>(std::round(string_cardinality_estimator.estimate()));
+    }
+
+    /// Cleans up any temporary or intermediate placeholder values.
+    void finalize() {
+      if (string_min_length == std::numeric_limits<size_t>::max()) {
+        string_min_length = 0;
+      }
+      if (string_min_length_not_empty == std::numeric_limits<size_t>::max()) {
+        string_min_length_not_empty = 0;
+      }
+    }
+
+    std::string column_name;
+
+    size_t rows_seen;
+    double numeric_mean;
+    size_t numeric_count;
+    size_t numeric_nan_count;
+    double numeric_min;
+    double numeric_max;
+
+    size_t string_min_length;
+    size_t string_min_length_not_empty;
+    size_t string_max_length;
+    size_t string_empty_count;
+    size_t string_only_whitespace_count;
+    size_t string_null_like_count;
+    std::unordered_set<std::string> string_captured_unique_values;
+    bool string_captured_unique_values_overflowed;
+    hll::HyperLogLog string_cardinality_estimator;
+    
+    std::string example_value;
+};
+
+class data_analysis {
```

Comment:
Can this be a simple aggregate instead of a class with a constructor. I don't see any logic in it. It simply holds a vector of columns. Why not simply declare it as:

```
struct data_analysis {
  std::vector<column_analysis> columns;
};
```